### PR TITLE
Substitutes variables into URL

### PR DESCRIPTION
**Pull request by jonkeller**
Previous work implemented ability for amp-analytics to create a 3p vendor iframe.
PR #11306 enables the iframe to send a response to the pub page
PR #11461 enables amp-ad-exit to receive those responses.
This PR enables amp-ad-exit to substitute them into target URLs.

### CODE REVIEW

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 1]
```diff
@@ -139,6 +156,13 @@ describes.realWin('amp-ad-exit', {
     win = env.win;
     toggleExperiment(win, 'amp-ad-exit', true);
     addAdDiv();
+    win.top.document.body.getResourceId = function() { return '6789'; };
```
Comment (glevitzky) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Any particular reason not to use `... = () => '6789';`?

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Nope...no reason other than I learned the "old way" and the "new way" isn't always instinctual yet. Will happily change it.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 2]
```diff
@@ -90,28 +93,72 @@ export class AmpAdExit extends AMP.BaseElement {
...
     };
     if (target.vars) {
-      for (const customVar in target.vars) {
-        if (customVar[0] == '_') {
-          vars[customVar] = () =>
-              args[customVar] == undefined ?
-                target.vars[customVar].defaultValue : args[customVar];
-          whitelist[customVar] = true;
+      for (const customVarName in target.vars) {
+        if (customVarName[0] == '_') {
```
Comment (keithwrightbos) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Consider inverting with continues to save indentation (makes code more readable IMO):
let customVar
if (customVarName[0] != '_' || !(customVar =  /** @type {!./config.Variable} */ (target.vars[customVarName]))) {
   continue;
}
substitutionFunctions[customVarName] = () => { ..

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 3]
```diff
@@ -90,28 +93,72 @@ export class AmpAdExit extends AMP.BaseElement {
...
+             The cross-domain iframe of vendorXYZ has sent the
+             following response for the creative:
+             { priority: medium, category: W }
+             This is just example data. The keys/values in that object can
+             be any strings.
+             The code below will create substitutionFunctions['_pty'],
+             which in this example will return "medium".
+            */
+            substitutionFunctions[customVarName] = () => {
+              if ('iframeTransportSignal' in customVar) {
```
Comment (keithwrightbos) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Same here... consider just returning if not in

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js [hunk 4]
```diff
@@ -90,28 +93,72 @@ export class AmpAdExit extends AMP.BaseElement {
...
+      for (const customVarName in target.vars) {
+        if (customVarName[0] == '_') {
+          const customVar =
+              /** @type {!./config.Variable} */ (target.vars[customVarName]);
+          if (customVar) {
+            /*
+             Example:
+             The amp-ad-exit target has a variable representing the
+             priority of something, which is defined as follows:
+             "vars": {
```
Comment (keithwrightbos) [inline extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Fix formatting

**Comment by jonkeller** [comment]
TODO: Rebase this onto 11461, so that testing can be done without manually copying files from one workspace to another.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 3 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Discussed verbally. Can invert, but would be wrong to just return if condition is false due to args and default value.

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 4 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Done

File: extensions/amp-ad-exit/0.1/amp-ad-exit.js (hunk 2 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/amp-ad-exit.js]: Done

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 5]
```diff
@@ -70,6 +73,20 @@ const EXIT_CONFIG = {
         },
       },
     },
+    variableFrom3pAnalytics: {
+      'finalUrl': 'http://localhost:8000/vars?foo=_foo',
+      vars: {
+        _foo: {
+          defaultValue: 'foo-default',
+          iframeTransportSignal:
+              'IFRAME_TRANSPORT_SIGNAL(' + TEST_3P_VENDOR +
```
Comment (zhouyx) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: nit: `IFRAME_TRANSPORT_SIGNAL(${TEST_3P_VENDOR}, collected-data)`

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 6]
```diff
@@ -139,6 +156,13 @@ describes.realWin('amp-ad-exit', {
     win = env.win;
     toggleExperiment(win, 'amp-ad-exit', true);
     addAdDiv();
+    win.top.document.body.getResourceId = () => '6789';
```
Comment (zhouyx) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: TODO: remove after rebase.

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js [hunk 7]
```diff
@@ -16,8 +16,11 @@
```
Comment (zhouyx) [inline extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: TODO: Add integration test : ) Please create separate issue

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Created https://github.com/ampproject/amphtml/issues/11503

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js (hunk 5 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Done

File: extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js (hunk 6 above)
Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Changed code to match other PR, added TODO comment

Comment (jonkeller) [reply extensions/amp-ad-exit/0.1/test/test-amp-ad-exit.js]: Actually, had to undo that. Will just add TODO as recommended.

**Pull request merged**